types/event_bus.go

```diff
@@ -90,52 +91,61 @@ func (b *EventBus) UnsubscribeAll(ctx context.Context, subscriber string) error
-	return nil
-}
-
-func (b *EventBus) validateAndStringifyTags(tags []cmn.KVPair, logger log.Logger) map[string]string {
-	result := make(map[string]string)
-	for _, tag := range tags {
-		// basic validation
-		if len(tag.Key) == 0 {
-			logger.Debug("Got tag with an empty key (skipping)", "tag", tag)
+	return b.pubsub.PublishWithEvents(ctx, eventData, map[string][]string{EventTypeKey: {eventType}})
+}
+
+// validateAndStringifyEvents takes a slice of event objects and creates a
+// map of stringified events where each key is composed of the event
+// type and each of the event's attributes keys in the form of
+// "{event.Type}.{attribute.Key}" and the value is each attribute's value.
+func (b *EventBus) validateAndStringifyEvents(events []types.Event, logger log.Logger) map[string][]string {
+	result := make(map[string][]string)
+	for _, event := range events {
+		if len(event.Type) == 0 {
+			logger.Debug("Got an event with an empty type (skipping)", "event", event)
 			continue
 		}
-		result[string(tag.Key)] = string(tag.Value)
+
+		for _, attr := range event.Attributes {
+			if len(attr.Key) == 0 {
+				logger.Debug("Got an event attribute with an empty key(skipping)", "event", event)
+				continue
+			}
+
+			compositeTag := fmt.Sprintf("%s.%s", event.Type, string(attr.Key))
+			result[compositeTag] = append(result[compositeTag], string(attr.Value))
+		}
 	}
+
 	return result
 }
 
 func (b *EventBus) PublishEventNewBlock(data EventDataNewBlock) error {
 	// no explicit deadline for publishing events
 	ctx := context.Background()
 
-	resultTags := append(data.ResultBeginBlock.Tags, data.ResultEndBlock.Tags...)
-	tags := b.validateAndStringifyTags(resultTags, b.Logger.With("block", data.Block.StringShort()))
+	resultEvents := append(data.ResultBeginBlock.Events, data.ResultEndBlock.Events...)
+	events := b.validateAndStringifyEvents(resultEvents, b.Logger.With("block", data.Block.StringShort()))
 
-	// add predefined tags
-	logIfTagExists(EventTypeKey, tags, b.Logger)
-	tags[EventTypeKey] = EventNewBlock
+	// add predefined new block event
+	events[EventTypeKey] = append(events[EventTypeKey], EventNewBlock)
 
-	b.pubsub.PublishWithTags(ctx, data, tags)
+	_ = b.pubsub.PublishWithEvents(ctx, data, events)
 	return nil
 }
 
 func (b *EventBus) PublishEventNewBlockHeader(data EventDataNewBlockHeader) error {
 	// no explicit deadline for publishing events
 	ctx := context.Background()
 
-	resultTags := append(data.ResultBeginBlock.Tags, data.ResultEndBlock.Tags...)
+	resultTags := append(data.ResultBeginBlock.Events, data.ResultEndBlock.Events...)
 	// TODO: Create StringShort method for Header and use it in logger.
-	tags := b.validateAndStringifyTags(resultTags, b.Logger.With("header", data.Header))
+	events := b.validateAndStringifyEvents(resultTags, b.Logger.With("header", data.Header))
 
-	// add predefined tags
-	logIfTagExists(EventTypeKey, tags, b.Logger)
```

Comment:
why remove?

Reply:
It was removed because the semantics changed.

Prior, we have had a `map[string]string` to represent tags/events. With this change set, we now have `map[string][]string`, so the notion of overwriting a key doesn't apply anymore as the value will simply just be appended.

Reply:
Missed this somehow, sorry.

> Maybe we can still warn that a Tendermint reserved duplicate will be added?

don't think there's a need.